livre/core/pipeline/PipeFilter.cpp

```diff
@@ -0,0 +1,235 @@
+    {
+        LBTHROW( std::runtime_error( std::string( "There is no port with name: ")
+                                                  + portName ));
+    }
+
+    void execute()
+    {
+        Futures inputFutures;
+        for( const auto& pair: _inputMap )
+        {
+            const Futures& futures = pair.second.getFutures();
+            for( const auto& future: futures )
+                inputFutures.push_back( future.rename( pair.second.getName( )));
+        }
+
+        const FutureMap futures( inputFutures );
+        PromiseMap promises( getOutputPromises( ));
+
+        try
+        {
+            _filter->execute( futures, promises );
+            promises.flush();
+        }
+        catch( const std::runtime_error& err )
+        {
+            promises.flush();
+            throw err;
+        }
+    }
+
+    Promise getInputPromise( const std::string& portName )
+    {
+      if( !hasInputPort( portName ))
+            throwPortError( portName );
+
+        if( _manuallySetPortsMap.count( portName ) > 0 )
+            return _manuallySetPortsMap.find( portName )->second.getPromise();
+
+        InputPort& inputPort = _inputMap.find( portName )->second;
+
+        _manuallySetPortsMap.emplace( std::piecewise_construct,
+                                      std::forward_as_tuple( inputPort.getName( )),
+                                      std::forward_as_tuple( DataInfo( inputPort.getName(),
+                                                                       inputPort.getDataType( ))));
+        OutputPort& outputPort = _manuallySetPortsMap.find( inputPort.getName( ))->second;
+
+        inputPort.connect( outputPort );
+        return _manuallySetPortsMap.find( portName )->second.getPromise();
+    }
+
+    Promises getOutputPromises() const
+    {
+        Promises promises;
+        for( const auto& pair: _outputMap )
+            promises.push_back( pair.second.getPromise( ));
+
+        return promises;
+    }
+
+    Futures getPostconditions() const
+    {
+        Futures futures;
+        for( const auto& pair: _outputMap )
+        {
+            const Future& outputFuture = pair.second.getPromise().getFuture();
+            futures.push_back( outputFuture );
+        }
+        return futures;
+    }
+
+    Futures getPreconditions() const
+    {
+        Futures futures;
+        for( const auto& pair: _inputMap )
+        {
+            const Futures& inputFutures = pair.second.getFutures();
+            futures.insert( futures.end(), inputFutures.begin(), inputFutures.end( ));
+        }
+        return futures;
+    }
+
+    void connect( const std::string& srcPortName,
+                  Impl& dstImpl,
+                  const std::string& dstPortName )
+    {
+        if( !hasOutputPort( srcPortName ))
+            throwPortError( srcPortName );
+
+        if( !dstImpl.hasInputPort( dstPortName ))
+            throwPortError( dstPortName );
```

Comment:
Not that this is going to be a bottleneck, but you could use find and here and compare to end() so you don't have to find again below.